### PR TITLE
feat: IONConnect DB conversation infrastructure

### PR DESCRIPTION
**Pull request by ice-damocles**
## Description
Adds 
- IONDatabaseNotifier to proxy DB from direct interaction
- General Event messages insert
- One-to-one conversation DB logic: init of the conversation, new messages, last message
- Group conversations DB logic: init of the conversation, new messages, change of subject, change of participants amount
- Test event messages mocks
- Changes to conversation provider: mocks are removed

## Type of Change
- [ ] Bug fix
- [x] New feature
- [x] Breaking change
- [ ] Refactoring
- [ ] Documentation
- [ ] Chore

### CODE REVIEW

File: lib/app/extensions/event_message.dart [hunk 1]
```diff
@@ -14,4 +14,18 @@ extension MasterKeyExtensions on EventMessage {
 
     return masterPubkey;
   }
+
+  String? get subject {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/extensions/event_message.dart:18]: We need to merge it to master first

File: lib/app/services/database/ion_database.c.dart [hunk 2]
```diff
@@ -23,34 +44,176 @@ class IONDatabase extends _$IONDatabase {
...
+    await _updateConversationMessagesTable(eventMessage);
+
+    return i;
+  }
+
+  // Inserts new conversation id or returns existing one
+  Future<void> _updateConversationMessagesTable(
+    EventMessage eventMessage,
+  ) async {
+    if (eventMessage.kind == 14) {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:113]: Done

File: lib/app/services/database/ion_database.c.dart [hunk 3]
```diff
@@ -23,34 +44,176 @@ class IONDatabase extends _$IONDatabase {
...
+            : conversationsWithChangedParticipants.first.conversationId;
+      }
+      return null;
+    }
+
+    return conversationsWithSameParticipants.first.conversationId;
+  }
+
+  Future<int> insertEventMessage(EventMessage eventMessage) async {
+    final i = await into(eventMessagesTable).insert(
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:99]: Now returns void

File: lib/app/extensions/event_message.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/extensions/event_message.dart:18]: Moved, extended `PrivateDirectMessageEntity` and added `RelatedSubject` to the `PrivateDirectMessageData`

File: lib/app/services/database/ion_database.c.dart [hunk 4]
```diff
@@ -23,34 +44,176 @@ class IONDatabase extends _$IONDatabase {
...
+    return i;
+  }
+
+  // Inserts new conversation id or returns existing one
+  Future<void> _updateConversationMessagesTable(
+    EventMessage eventMessage,
+  ) async {
+    if (eventMessage.kind == 14) {
+      // This is the first message of the one-to-one conversation OR
+      // change of the group conversation subject
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:115]: Will fix and write unit tests for it

File: lib/app/features/chat/recent_chats/providers/conversations_provider.c.dart [hunk 5]
```diff
@@ -15,17 +19,81 @@ class Conversations extends _$Conversations {
...
+        .read(iONDatabaseNotifierProvider.notifier)
+        .watchConversations()
+        .listen((conversationsEventMessages) async {
+      final data = await _mapConversations(conversationsEventMessages);
+
+      state = AsyncValue.data(data);
+    });
+  }
+
+  Future<List<RecentChatDataModel>> _mapConversations(
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/recent_chats/providers/conversations_provider.c.dart:45]: Refactored `ConversationProvider` now uses and returns `PrivateDirectMessageEntity`, UI will be updated as part of another task

File: lib/app/services/database/ion_database.c.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:115]: Added unit tests

[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:115]: Added unit tests and fixed logic
image

File: lib/app/features/feed/data/models/entities/related_subject.c.dart
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/feed/data/models/entities/related_subject.c.dart:1]: Moved

File: lib/app/features/chat/views/pages/chat_main_page/chat_main_page.dart [hunk 6]
```diff
@@ -9,7 +9,7 @@ import 'package:ion/app/features/chat/recent_chats/views/pages/recent_chats_empt
 import 'package:ion/app/features/chat/recent_chats/views/pages/recent_chats_timeline_page/recent_chats_timeline_page.dart';
 import 'package:ion/app/features/chat/views/pages/chat_main_page/components/chat_main_appbar/chat_main_appbar.dart';
 
-class ChatMainPage extends ConsumerWidget {
+class ChatMainPage extends HookConsumerWidget {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/views/pages/chat_main_page/chat_main_page.dart:12]: Removed

File: lib/app/services/database/ion_database.c.dart [hunk 7]
```diff
@@ -4,53 +4,234 @@ import 'dart:convert';
...
+
+    await _updateConversationMessagesTable(eventMessage);
+  }
+
+  // Inserts new conversation id or returns existing one
+  Future<void> _updateConversationMessagesTable(
+    EventMessage eventMessage,
+  ) async {
+    if (eventMessage.kind == PrivateDirectMessageEntity.kind) {
+      // This is the first message of the one-to-one conversation OR
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:85]: Removed

File: lib/app/services/database/ion_database.c.dart [hunk 8]
```diff
@@ -4,53 +4,234 @@ import 'dart:convert';
...
+    final lastConversationEventMessages = (await (select(eventMessagesTable)
+              ..where((table) => table.id.isIn(lastConversationMessagesIds)))
+            .get())
+        .map((e) => e.toEventMessage())
+        .toList();
+
+    return lastConversationEventMessages;
+  }
+
+  Stream<List<EventMessage>> watchConversations() {
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:172]: `getAllConversations` doesn't need to wait for DB update, it just fetches all conversation present in the table
`watchConversations().first` would return empty as it is waiting for the table update which could not happen

File: lib/app/services/database/ion_database.c.dart [hunk 9]
```diff
@@ -4,53 +4,234 @@ import 'dart:convert';
...
+  Future<void> insertEventMessage(EventMessage eventMessage) async {
+    await into(eventMessagesTable).insert(
+      EventMessageTableData.fromEventMessage(eventMessage),
+      mode: InsertMode.insertOrReplace,
+    );
+
+    await _updateConversationMessagesTable(eventMessage);
+  }
+
+  // Inserts new conversation id or returns existing one
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:80]: Removed

File: lib/app/features/chat/model/entities/private_direct_message_data.c.dart [hunk 10]
```diff
@@ -27,8 +28,9 @@ class PrivateDirectMessageEntity with _$PrivateDirectMessageEntity {
 
   const PrivateDirectMessageEntity._();
 
-  /// https://github.com/nostr-protocol/nips/blob/master/17.md
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/features/chat/model/entities/private_direct_message_data.c.dart:30]: Responded in Slack

File: lib/app/services/database/ion_database.c.dart [hunk 11]
```diff
@@ -4,53 +4,239 @@ import 'dart:convert';
...
+      final conversationIdByPubkeys = await _lookupConversationByPubkeys(conversationMessage);
+
+      // Existing conversation (one-to-one or group)
+      if (conversationIdByPubkeys != null) {
+        await insertConversationData(
+          eventMessage: eventMessage,
+          conversationId: conversationIdByPubkeys,
+        );
+      }
+      // Existing group conversation (change of participants)
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:94]: You mean wrong place? Cause if we search by subject it means amount of participants has changed

File: lib/app/services/database/ion_database.c.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:176]: `customSelectStream` is deprecated and also it serves same purpose as `.watch` which returns stream but only during the updates, it doesn't fetch data initially

File: lib/app/services/database/ion_database.c.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/services/database/ion_database.c.dart:94]: Fixed

File: pubspec.yaml [hunk 12]
```diff
@@ -156,7 +157,6 @@ flutter:
     - assets/images/notifications/
     - assets/emoji_data_sets/emojis-by-group.json
     - assets/emoji_data_sets/emoji-keywords.json
-    - assets/ion_identity/coins.json
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply pubspec.yaml:159]: It is not used and I get lint warning

File: lib/app/exceptions/exceptions.dart [hunk 13]
```diff
@@ -164,6 +164,10 @@ class DecompressBrotliException extends IONException {
   DecompressBrotliException() : super(10034, 'Failed to decompress Brotli file.');
 }
 
+class ConversationIsNotFoundException extends IONException {
+  ConversationIsNotFoundException() : super(10035, 'Failed to find conversation');
```
[the remark this answers is not in the document]
Comment (ice-damocles) [reply lib/app/exceptions/exceptions.dart:168]: Fixed all

**Review by ice-kreios — APPROVED** [review]

**Pull request merged**